Free surface tractions

fionaclerc opened the pull request:
This pull request add an exception to allow boundary tractions to be applied on the free surface. It also includes benchmarks for the surface deformation of viscous and viscoelastic materials in response to surface loading.

naliboff commented on benchmarks/free_surface_tractions/viscous/free_surface_viscous_cylinder_2D_loading.prm:
```diff
@@ -0,0 +1,159 @@
+
+# Mesh refinement specifications 
+subsection Mesh refinement
+  set Initial adaptive refinement        = 1
```
Do the results change significantly if you do not do the adaptive refinement? If so, it might be worth putting in a brief comment mentioning this.

naliboff commented on benchmarks/free_surface_tractions/viscous/free_surface_viscous_cylinder_2D_loading.prm:
```diff
@@ -0,0 +1,159 @@
+        # r0 is load radius, H0 is load height, t1 is time load is removed, 
+	# rhoi is density of ice/load, g is gravitational acceleration 
+    set Function expression = 0; if (x<r0 && t<t1,-g*H0*rhoi*(1-t/t1), 0)
+    #set Function expression = 0; if (x<r0 ,if(t<t0,-g*H0*rhoi,if(t<t1,-g*H0*rhoi*(1-(t-t0)/t1),0)), 0)
```
Remove this line.

naliboff commented on benchmarks/free_surface_tractions/viscous/free_surface_viscous_cylinder_2D_loading.prm:
```diff
@@ -0,0 +1,159 @@
+
+# Composition boundary conditions
+subsection Boundary composition model
+  set Allow fixed composition on outflow boundaries = true
```
Do you need this line if there are no` Fixed composition boundary indicators`?

naliboff commented on benchmarks/free_surface_tractions/viscous/free_surface_viscous_cylinder_2D_loading.prm:
```diff
@@ -0,0 +1,159 @@
+# Temperature boundary conditions
+subsection Boundary temperature model
+  set Fixed temperature boundary indicators = bottom, top, left, right
+  set List of model names = box
```
You can change this line to ` set List of model names = initial temperature`, which would allow deleting the `Box` subsection below.

naliboff commented on benchmarks/free_surface_tractions/viscous/free_surface_viscous_cylinder_3D_loading.prm:
```diff
@@ -0,0 +1,162 @@
+# Material model
+subsection Material model
+
+   set Model name = simple
```
For simplicity, you could use the `viscoelastic` material model here and set the `Elastic shear moduli` to something like 1e50, which for all intensive purposes produces viscous behavior (e.g., negligible viscoelastic contribution).

naliboff commented on benchmarks/free_surface_tractions/viscous/free_surface_viscous_cylinder_3D_loading.prm:
```diff
@@ -0,0 +1,162 @@
+  end
+end
+
+# Number and name of compositional fields
```
Unless you use the `Viscoelastic` material model below, there is no need to specify these compositional fields.

naliboff commented on benchmarks/free_surface_tractions/viscous/free_surface_viscous_cylinder_3D_loading.prm:
```diff
@@ -0,0 +1,162 @@
+# Temperature boundary conditions
+subsection Boundary temperature model
+  set Fixed temperature boundary indicators = bottom, top, left, right, front, back
+  set List of model names = box
```
```suggestion
  set List of model names = initial temperature
```

and then delete the `Box` subsection.

naliboff commented on source/simulator/helper_functions.cc:
```diff
@@ -2040,7 +2040,7 @@ namespace aspect
       = { boundary_velocity_manager.get_zero_boundary_velocity_indicators(),
           boundary_velocity_manager.get_tangential_boundary_velocity_indicators(),
-          parameters.free_surface_boundary_indicators,
+      //    parameters.free_surface_boundary_indicators,
```
Please delete this line.

naliboff commented:
/rebuild

gassmoeller commented:
@fionaclerc do you want to talk about how to address the comments?

naliboff commented:
@fionaclerc - is this ready for another review?

fionaclerc commented on benchmarks/free_surface_tractions/viscoelastic/README:
```diff
@@ -0,0 +1,39 @@
+Changing the stress instantaneously as the load is applied/removed
+ leads to oscillations in the elastic response. This can be addressed
+ by setting a fixed elastic timestep larger than the numerical 
+ timestep and turning on stress averaging:
```
Using a smaller CFL value does not reduce the oscillations. Setting the maximum time step equal to the elastic time step does work, as in the parameter files “free_surface_VE_cylinder_*D_loading_fixed_elastic_dt.prm”. I am correcting/clarifying the README file.

fionaclerc commented on benchmarks/free_surface_tractions/viscoelastic/README:
```diff
@@ -0,0 +1,39 @@
+ r0 is the load radius), for t>0. The load is fully removed by t=t1,
+ and the surface relaxes. This is done both in a 2-D and 3-D geometry
+ (by symmetry the load is centered on the left boundary or left/front
+ corner). The input files are:
```
Yes.

fionaclerc commented:
I have updated the README files and some comments to address the review. I also added a note in the README files about the mismatch between the analytical and numerical solutions at the far (right) boundary due to the free-slip condition.

tjhei commented:
thanks for continuing the work on this! Sadly, there have been edits in free_surface.cc merged since you started your branch. You will need to rebase your changes to the current master. Note that the free_surface.cc got moved into a different folder, so you likely have to fix the conflicts manually. Let us know if you need help.

naliboff commented:
@fionaclerc and I are working on fixing the rebase, will rebuild and test after that is done.

naliboff commented:
@fionaclerc - This looks great, well done and thanks for sticking through all of the revisions! These benchmarks will be incredibly useful for a large group of users.

The last step is to do a final commit adding a changelog entry in the  doc/modules/changes directory. A good example for this type of addition is: doc/modules/changes/20190529_euen. 

Let me know if you have any questions about this and as soon the change log file is added, this will be ready to merge.
